output.go

```diff
@@ -0,0 +1,198 @@
+}
+
+const schema = `
+	CREATE TABLE IF NOT EXISTS samples (
+		ts timestamptz NOT NULL DEFAULT current_timestamp,
+		metric varchar(128) NOT NULL,
+		tags jsonb,
+		value real
+	);
+	CREATE TABLE IF NOT EXISTS thresholds (
+		id serial,
+		ts timestamptz NOT NULL DEFAULT current_timestamp,
+		metric varchar(128) NOT NULL,
+		threshold varchar(128) NOT NULL,
+		abort_on_fail boolean DEFAULT FALSE,
+		delay_abort_eval varchar(128),
+		last_failed boolean DEFAULT FALSE
+	);
+	SELECT create_hypertable('samples', 'ts');
+	CREATE INDEX IF NOT EXISTS idx_samples_ts ON samples (ts DESC);
+	CREATE INDEX IF NOT EXISTS idx_thresholds_ts ON thresholds (ts DESC);`
+
+func (o *Output) Start() error {
+	conn, err := o.Pool.Acquire(context.Background())
+	if err != nil {
+		o.logger.WithError(err).Error("Start: Couldn't acquire connection")
+	}
+	defer conn.Release()
+
+	_, err = conn.Exec(context.Background(), "CREATE DATABASE "+o.Config.dbName.String)
+	if err != nil {
+		o.logger.WithError(err).Debug("Start: Couldn't create database; most likely harmless")
+	}
+	_, err = conn.Exec(context.Background(), schema)
+	if err != nil {
+		o.logger.WithError(err).Debug("Start: Couldn't create database schema; most likely harmless")
+	}
+
+	for metric, thresholds := range o.thresholds {
+		for _, t := range thresholds {
+			err = conn.QueryRow(context.Background(), `
+				INSERT INTO thresholds (metric, threshold, abort_on_fail, delay_abort_eval, last_failed)
+				VALUES ($1, $2, $3, $4, $5)
+				RETURNING id`,
+				metric, t.threshold.Source, t.threshold.AbortOnFail, t.threshold.AbortGracePeriod.String(), t.threshold.LastFailed).
+				Scan(&t.id)
+			if err != nil {
+				o.logger.WithError(err).Debug("Start: Failed to insert threshold")
+			}
+		}
+	}
+
+	pf, err := output.NewPeriodicFlusher(time.Duration(o.Config.PushInterval.Duration), o.flushMetrics)
+	if err != nil {
+		return err
+	}
+
+	o.logger.Debug("Start: Running!")
+	o.periodicFlusher = pf
+
+	return nil
+}
+
+func (o *Output) flushMetrics() {
+	sampleContainers := o.GetBufferedSamples()
+	start := time.Now()
+
+	var batch pgx.Batch
+	for _, sc := range sampleContainers {
+		samples := sc.GetSamples()
+		o.logger.Debug("flushMetrics: Committing...")
+		o.logger.WithField("samples", len(samples)).Debug("flushMetrics: Writing...")
+
+		for _, s := range samples {
+			tags := s.Tags.CloneTags()
+			batch.Queue(`INSERT INTO samples (ts, metric, value, tags) VALUES ($1, $2, $3, $4)`,
+				s.Time, s.Metric.Name, s.Value, tags)
+		}
+	}
+
+	for _, t := range o.thresholds {
+		for _, threshold := range t {
+			batch.Queue(`UPDATE thresholds SET last_failed = $1 WHERE id = $2`,
+				threshold.threshold.LastFailed, threshold.id)
+		}
+	}
+
+	conn, err := o.Pool.Acquire(context.Background())
+	if err != nil {
+		o.logger.WithError(err).Error("flushMetrics: Couldn't acquire connection to write samples")
+		return
+	}
+	defer conn.Release()
+
+	br := conn.SendBatch(context.Background(), &batch)
```

Comment:
From the [documentation](https://pkg.go.dev/github.com/jackc/pgx/v4#Conn.SendBatch) it seems like we should close the batch?